cpp/src/stream_compaction/drop_duplicates.cu

```diff
@@ -0,0 +1,303 @@
+                                              nulls_are_equal);
+    return thrust::count_if(rmm::exec_policy(stream)->on(stream),
+              thrust::counting_iterator<cudf::size_type>(0),
+              thrust::counting_iterator<cudf::size_type>(keys.num_rows()),
+              [sorted_row_index, comp]
+              __device__ (cudf::size_type i) mutable {
+              return (i == 0 || not comp(sorted_row_index[i], sorted_row_index[i-1]));
+              });
+  }
+}
+
+std::unique_ptr<experimental::table>
+  drop_duplicates(table_view const& input,
+                  table_view const& keys,
+                  duplicate_keep_option const& keep,
+                  bool const& nulls_are_equal,
+                  rmm::mr::device_memory_resource* mr,
+                  cudaStream_t stream)
+{
+  if (0 == input.num_rows() || 
+      0 == input.num_columns() ||
+      0 == keys.num_columns()
+      ) {
+      std::vector<std::unique_ptr<column>> out_columns(input.num_columns());
+      std::transform(input.begin(), input.end(), out_columns.begin(),
+                [&stream] (auto col_view){
+                return detail::empty_like(col_view, stream);
+                });
+
+    return std::make_unique<experimental::table>(std::move(out_columns));
+  }
+  
+  CUDF_EXPECTS( input.num_rows() == keys.num_rows(), "number of \
+rows in input table should be equal to number of rows in key colums table");
+
+  // The values will be filled into this column
+  auto unique_indices = 
+        cudf::make_numeric_column(data_type{INT32}, 
+                                  keys.num_rows(), UNALLOCATED, stream, mr);
+  // This is just slice of `unique_indices` but with different size as per the
+  // keys has been processed in `get_unique_ordered_indices`
+  auto unique_indices_view = 
+      detail::get_unique_ordered_indices(keys, 
+                                         *unique_indices,
+                                         keep, nulls_are_equal);
+ 
+  // run gather operation to establish new order
+  return detail::gather(input, unique_indices_view, false, false, true, mr, stream);
+}
+
+template <typename T>
+struct check_for_nan
+{
+    check_for_nan(cudf::column_device_view input) :_input{input}{}
+  __device__
+  bool operator()(size_type index)
+  {
+    return std::isnan(_input.data<T>()[index]) and _input.is_valid(index);
+  }
+
+protected:
+  cudf::column_device_view _input;
+};
+
+cudf::size_type unique_count(column_view const& input,
+                             bool const& ignore_nulls,
+                             bool const& nan_as_null,
+                             rmm::mr::device_memory_resource *mr,
+                             cudaStream_t stream)
+{
+  if (0 == input.size() || input.null_count() == input.size()) {
+    return 0;
+  }
+
+  cudf::size_type nrows = input.size();
+ 
+  bool has_nans = false;
+  // Check for Nans
+  if (input.type().id() == FLOAT32 and input.has_nulls() and nan_as_null) {
```

Comment:
To avoid duplication between the float32 and float64, I'd make this logic into a type-dispatched function specialized for `std::is_floating_point<T>::value`. For `not std::is_floating_point<T>::value`, just return false.